File: Recursion/lecture 9/Fraz_recursion_9.cpp
#include <bits/stdc++.h>
void perm(int pos, string &str, vector<string> &ans)
{
    if (pos >= str.size())
    {
        ans.push_back(str);
    }
    unordered_set<char> m;

    for (int i = pos; i < str.size(); i++)
    {
        if (m.find(str[i]) != m.end())
            continue;
        m.insert(str[i]);
        swap(str[i], str[pos]);
        perm(pos + 1, str, ans);
        swap(str[i], str[pos]); // backtracking
    }
}

vector<string> generatePermutations(string &str)
{
    vector<string> ans;
    perm(0, str, ans);
    sort(ans.begin(), ans.end());
    return ans;
}

File: Recursion/lecture 9/Fraz_recursion_9.java
public static HashSet<String> h = new HashSet<String>();
    ArrayList<String> ans = new ArrayList<>();

    public void permute(char s[], int i, int n) {

        // If the permutation is complete
        if (i == n) {

            // If set doesn't contain
            // the permutation already
            if (!(h.contains(String.copyValueOf(s)))) {

                h.add(String.copyValueOf(s));

                ans.add(String.copyValueOf(s));
            }
        }

        else {

            // One by one swap the jth
            // character with the ith
            for (int j = i; j <= n; j++) {

                // Swapping a[i] and a[j];
                char temp = s[i];
                s[i] = s[j];
                s[j] = temp;

                // Revert the swapping or backtracking
                permute(s, i + 1, n);

                temp = s[i];
                s[i] = s[j];
                s[j] = temp;
            }
        }
    }

    ArrayList<String> generatePermutations(String str) {

        char[] ch = str.toCharArray();
        permute(ch, 0, ch.length - 1);
        Collections.sort(ans);
        return ans;
    }